Student: Arora, Sanjana V00966221

Question 11

In [ ]:
import numpy as np
import pandas as pd
import sklearn
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

In [ ]:
!wget "https://raw.githubusercontent.com/fivethirtyeight/data/master/candy-power-ranking/candy-data.csv"
df = pd.read_csv("candy-data.csv")

--2021-06-07 02:38:42--  https://raw.githubusercontent.com/fivethirtyeight/data/master/candy-power-ranking/candy-data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5205 (5.1K) [text/plain]
Saving to: ‘candy-data.csv’

candy-data.csv      100%[===================>]   5.08K  --.-KB/s    in 0s      

2021-06-07 02:38:42 (54.0 MB/s) - ‘candy-data.csv’ saved [5205/5205]



In [ ]:
df.head()

competitorname  chocolate  fruity  ...  sugarpercent  pricepercent  winpercent
0      100 Grand          1       0  ...         0.732         0.860   66.971725
1   3 Musketeers          1       0  ...         0.604         0.511   67.602936
2       One dime          0       0  ...         0.011         0.116   32.261086
3    One quarter          0       0  ...         0.011         0.511   46.116505
4      Air Heads          0       1  ...         0.906         0.511   52.341465

[5 rows x 13 columns]

Trying to predict whether candy is a "chocolate" or not using linear regression. So trying to predict value of chocolate.

In [ ]:
df.columns

Index(['competitorname', 'chocolate', 'fruity', 'caramel', 'peanutyalmondy',
       'nougat', 'crispedricewafer', 'hard', 'bar', 'pluribus', 'sugarpercent',
       'pricepercent', 'winpercent'],
      dtype='object')

In [ ]:
Subset = df[['fruity', 'caramel', 'peanutyalmondy',
       'nougat', 'crispedricewafer', 'hard', 'bar', 'pluribus', 'sugarpercent',
       'pricepercent', 'winpercent']].copy()

In [ ]:
Y= df['chocolate']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(Subset, Y, test_size=0.3,shuffle = True)

In [ ]:
clf = LinearRegression()

In [ ]:
clf.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
Y_pred = clf.predict(X_train)

In [ ]:
#Training Mean Absolute Error
print('The training MAE:', mean_absolute_error(Y_train, Y_pred))

The training MAE: 0.17113453101704634


In [ ]:
Y_pred_test = clf.predict(X_test)

In [ ]:
#Testing Mean Absolute Error
print('The testing MAE:', mean_absolute_error(Y_test, Y_pred_test))

The testing MAE: 0.24054777757907725


We can use linear regression for binary classification. However, the training and testing mean absolute errors are very high.This indicates that linear regression is not best suited for binary classification. Linear regression is usually better for continuous target features. 


Logistic regression is generally considered to be a better machine learning model for predicting an outcome in binary classification due to the powerful transformation of the input data using the sigmoid function. 

**Implementing Gradient Descent Based Logistic Regression**
1. Splitting the dataset
2. Determining independent features and target feature
3. Calculating Sigmoid
4. Logistic Regression Classifier Function for calculating weights and intercept of the logistic regression
5. Fit function for generating the predicted values

In [ ]:
def train_validation_split(dataset):
    validation_split = .3
    dataset_size = len(dataset)
    split = int(np.floor(validation_split * dataset_size))
    training_data = dataset.iloc[split:].reset_index(drop=True)
    validation_data = dataset.iloc[:split].reset_index(drop=True)
    return training_data,validation_data

In [ ]:
Subset1 = df[['fruity', 'caramel', 'peanutyalmondy',
       'nougat', 'crispedricewafer', 'hard', 'pluribus', 'sugarpercent',
       'pricepercent', 'winpercent','chocolate']].copy()

In [ ]:
Subset1 = Subset1.sample(frac = 1)
train_data = train_validation_split(Subset1)[0]
test_data = train_validation_split(Subset1)[1]

In [ ]:
X_train = train_data.iloc[:,:-1]
Y_train = train_data['chocolate']

In [ ]:
X_test = test_data.iloc[:,:-1].values
Y_test = test_data['chocolate']

**Logistic Regression from scratch**

In [ ]:
# Calculating sigmoid function
def sigmoid(outcome):
    return 1 / (1 + np.exp(-outcome))

In [ ]:
def log_reg(X, Y, steps, learning_rate):
    
    # calculating intercept
    intercept = np.ones((len(X), 1))
    X = np.hstack((intercept, X))
      # initializing weights  
    weights = np.zeros(X.shape[1])
    
    for i in range(steps):
        outcome = np.dot(X, weights)
        predictions = sigmoid(outcome)
       
        #calculating gradient
        gradient = np.dot(X.T, Y - predictions)
        # Update weights with gradient
        weights += learning_rate * gradient
    return weights

In [ ]:
#This function is used for predicting the classification result from the Logistic Regression
def fit(X,Weights):

    #matrix storing the X_input and Intercept
    X_and_intercept_matrix = np.hstack((np.ones((len(X), 1)),X))
    
    # predictions                                   
    preds = np.round(sigmoid(np.dot(X_and_intercept_matrix, Weights)))
    
    return preds

In [ ]:
# Training the model
calculate_weights = log_reg(X_train,Y_train,steps = 15000, learning_rate = 0.000001)
                     

In [ ]:
# generating predicted values
Y_predlr= fit(X_train,calculate_weights)

In [ ]:
Y_predtestlr=fit(X_test,calculate_weights)

In [ ]:
# Training Mean Absolute Error
np.mean(np.square(np.array(Y_predlr)-np.array(Y_train)))

0.13333333333333333

In [ ]:
#Testing Mean Absolute Error
np.mean(np.square(np.array(Y_predtestlr)-np.array(Y_test)))

0.2

The training (0.133) and testing error(0.2) after using logistic regression are better than the training(0.17) and testing errors derived from(0.24) linear regression model.

In [ ]:
#Referred from https://towardsdatascience.com/logistic-regression-from-scratch-69db4f587e17